File: python/refrase/engine.py
"""Rule engine -- reads a family config and applies matching transforms in sequence."""

from typing import Any, Optional

from refrase.transforms import TRANSFORMS


def should_apply(when: dict[str, Any], variant: str, task: str) -> bool:
    """Check if a rule's conditions match the current variant and task."""
    variants = when.get("variants") or []
    variant_match = (
        len(variants) == 0
        or "all" in variants
        or variant in variants
    )
    tasks = when.get("tasks") or []
    task_match = (
        len(tasks) == 0
        or "all" in tasks
        or task in tasks
    )
    return variant_match and task_match


def apply_rules(
    config: dict[str, Any],
    model_id: str,
    variant: str,
    task: str,
    system_prompt: str,
    user_prompt: Optional[str] = None,
) -> dict[str, Any]:
    """Apply all matching rules from a family config to produce an AdaptResult dict.

    Returns:
        Dict with keys: system, user, model_id, model_family, changes, and
        optionally api_hints.
    """
    system = system_prompt
    user: Optional[str] = user_prompt
    changes: list[dict[str, Any]] = []
    api_hints: list[dict[str, Any]] = []

    for rule in config["rules"]:
        if not should_apply(rule["when"], variant, task):
            continue

        transform = TRANSFORMS.get(rule["transform"])
        if transform is None:
            raise ValueError(
                f'Unknown transform "{rule["transform"]}" in rule '
                f'"{rule["id"]}" for family "{config["family"]}"'
            )

        # Build params: merge rule params with engine context
        params: dict[str, Any] = {
            **(rule.get("params") or {}),
            "variant": variant,
            "task": task,
        }

        target = rule["target"]
        if target == "system" or target == "both":
            system = transform(system, params)
        if (target == "user" or target == "both") and user is not None:
            user = transform(user, params)

        evidence = rule.get("evidence") or {}
        changes.append({
            "rule": rule["id"],
            "description": rule["description"],
            "evidence": evidence.get("source", ""),
            "impact": rule["impact"],
            "category": rule.get("category"),
        })

    # Collect matching API hints
    for hint in config.get("api_hints") or []:
        hint_when = hint.get("when")
        if hint_when is None or should_apply(hint_when, variant, task):
            api_hints.append({
                "parameter": hint["parameter"],
                "value": hint["value"],
                "reason": hint["reason"],
                "evidence": hint.get("evidence"),
            })

    result: dict[str, Any] = {
        "system": system,
        "user": user,
        "model_id": model_id,
        "model_family": config["family"],
        "changes": changes,
    }

    if api_hints:
        result["api_hints"] = api_hints

    return result

File: typescript/src/engine.ts
/**
 * Rule engine — reads a family config and applies matching transforms in sequence.
 */
import { TRANSFORMS } from "./transforms/index.js";
import type {
  AdaptResult,
  ApiHint,
  Change,
  FamilyConfig,
  ModelFamily,
  RuleConfig,
  RuleWhen,
  TaskType,
} from "./types.js";

/**
 * Check if a rule's conditions match the current variant and task.
 */
function shouldApply(when: RuleWhen, variant: string, task: TaskType): boolean {
  const variantMatch =
    !when.variants ||
    when.variants.length === 0 ||
    when.variants.includes("all") ||
    when.variants.includes(variant);
  const taskMatch =
    !when.tasks ||
    when.tasks.length === 0 ||
    when.tasks.includes("all") ||
    when.tasks.includes(task);
  return variantMatch && taskMatch;
}

/**
 * Apply all matching rules from a family config to produce an AdaptResult.
 */
export function applyRules(
  config: FamilyConfig,
  modelId: string,
  variant: string,
  task: TaskType,
  systemPrompt: string,
  userPrompt?: string,
): AdaptResult {
  let system = systemPrompt;
  let user: string | null = userPrompt ?? null;
  const changes: Change[] = [];
  const apiHints: ApiHint[] = [];

  for (const rule of config.rules) {
    if (!shouldApply(rule.when, variant, task)) continue;

    const transform = TRANSFORMS[rule.transform];
    if (!transform) {
      throw new Error(
        `Unknown transform "${rule.transform}" in rule "${rule.id}" for family "${config.family}"`,
      );
    }

    // Build params: merge rule params with engine context
    const params: Record<string, unknown> = {
      ...rule.params,
      variant,
      task,
    };

    if (rule.target === "system" || rule.target === "both") {
      system = transform(system, params);
    }
    if ((rule.target === "user" || rule.target === "both") && user !== null) {
      user = transform(user, params);
    }

    changes.push({
      rule: rule.id,
      description: rule.description,
      evidence: rule.evidence?.source ?? "",
      impact: rule.impact,
      category: rule.category,
    });
  }

  // Collect matching API hints
  for (const hint of config.api_hints ?? []) {
    if (!hint.when || shouldApply(hint.when, variant, task)) {
      apiHints.push({
        parameter: hint.parameter,
        value: hint.value,
        reason: hint.reason,
        evidence: hint.evidence,
      });
    }
  }

  const result: AdaptResult = {
    system,
    user,
    modelId,
    modelFamily: config.family as ModelFamily,
    changes,
  };

  if (apiHints.length > 0) {
    result.apiHints = apiHints;
  }

  return result;
}
